client/beefy/rpc/src/lib.rs

```diff
@@ -20,95 +20,78 @@
 
 #![warn(missing_docs)]
 
-use std::sync::Arc;
-
-use sp_runtime::traits::Block as BlockT;
-
-use futures::{FutureExt, SinkExt, StreamExt};
-use jsonrpc_derive::rpc;
-use jsonrpc_pubsub::{manager::SubscriptionManager, typed::Subscriber, SubscriptionId};
-use log::warn;
-
 use beefy_gadget::notification::BeefySignedCommitmentStream;
+use futures::{future, task::Spawn, FutureExt, StreamExt};
+use jsonrpsee::{
+	proc_macros::rpc,
+	types::{Error as JsonRpseeError, RpcResult},
+	SubscriptionSink,
+};
+use log::warn;
+use sc_rpc::SubscriptionTaskExecutor;
+use sp_runtime::traits::Block as BlockT;
 
 mod notification;
 
 /// Provides RPC methods for interacting with BEEFY.
-#[rpc]
+#[rpc(client, server, namespace = "beefy")]
 pub trait BeefyApi<Notification, Hash> {
-	/// RPC Metadata
-	type Metadata;
-
 	/// Returns the block most recently finalized by BEEFY, alongside side its justification.
-	#[pubsub(
-		subscription = "beefy_justifications",
-		subscribe,
-		name = "beefy_subscribeJustifications"
+	#[subscription(
+		name = "subscribeJustifications",
+		aliases = ["beefy_justifications"],
```

Comment:
ah ok I never got this clearly, I think I was confused by this and hacked around just to make polkadot-js happy.

correct, yes but the notification_method will be `beef_subscribeJustifications`.

Reply:
Hmm, then it's a bit of a breaking change.

We should be subscribing with: `beefy_subscribeJustifications` and then the client should start receiving `beefy_justifications` notifications. Anyway to express this in jsonrpsee?

Reply:
no not really currently, should be quite easy to fix though...

what's the main motivation for subscribing to a method that sends notifications on method with a different name BTW?

Reply:
> Hmm, then it's a bit of a breaking change.

If we can get away with it, I would prefer to break backwards compatibility. It's a bit of a weird feature and I'm not sure why it is useful (but maybe I'm missing something).

@tomusdrw Are you aware of any instance in the wild that is relying on subscribing with one method name and getting the responses under a different name?

Reply:
> It's a bit of a weird feature and I'm not sure why it is useful (but maybe I'm missing something).

Not sure I get the point. The way subscriptions work is that both ends act as both client & server. When you subscribe to something you instruct the server to start sending `notifications` to the other end (i.e. act as a client and treat the other end like a server that can receive RPC calls as well). It seems reasonable for the notification to use a different method name than the one used for subscribing to easily distinguish the two.
In Ethereum that's how it was designed, hence we needed the distinction, but if it makes it simpler to implement in `jsonrpsee/substrate` and/or you feel this is more reasonable, I'm fine with that.

Reply:
Thanks for the context.

You mean to distinguish between `method call "subscribeFoo"` and `notification "subscribeFoo"` correct (not to distinguish server vs client notifications)?

We distinguish the two because notifications and requests are different types so I don't think a different name is "necessary" rather than that it's more human-friendly.... but I suppose it depends on how other libraries deals with especially in JS world..

am I misunderstanding something?

//cc @maciejhirsz :monocle_face:

Reply:
It's rather easy to implement so why not support it for now to avoid breaking something?

Reply:
> The way subscriptions work is that both ends act as both client & server.

Ok, when you frame it like that it makes more sense. I still think it's of very limited usefulness to allow users to name things arbitrarily. Clients only ever send `subscribeFoo`/`unsubscribeFoo` calls, so there's little chance of ambiguity on the client side as it only **receives** responses named `subscribeFoo`. Sane client side code would only ever use the subscription ID to distinguish responses so yeah, unless we know we need this I'd prefer we keep it out.

Reply:
@tomusdrw we had a conversation about this and decided to support the same semantics `jsonrpc` have here. Not loving it, but best not upset user land code. :)

Reply:
To hammer home the point, we necessarily need it for frontier and moonbeam, as we are exposing RPC subscriptions that must be usable by the existing tools of the ethereum eco-system as is.

Reply:
Addressed in https://github.com/paritytech/jsonrpsee/pull/568